lightning/src/routing/router.rs

```diff
@@ -333,84 +671,298 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
+						if let Some(node_info) = node.announcement_info.as_ref() {
+							ordered_hops.last_mut().unwrap().route_hop.node_features = node_info.features.clone();
+						} else {
+							ordered_hops.last_mut().unwrap().route_hop.node_features = NodeFeatures::empty();
+						}
 					} else {
-						res.last_mut().unwrap().node_features = NodeFeatures::empty();
+						// We should be able to fill in features for everything except the last
+						// hop, if the last hop was provided via a BOLT 11 invoice (though we
+						// should be able to extend it further as BOLT 11 does have feature
+						// flags for the last hop node itself).
+						assert!(ordered_hops.last().unwrap().route_hop.pubkey == *payee);
 					}
-				} else {
-					// We should be able to fill in features for everything except the last
-					// hop, if the last hop was provided via a BOLT 11 invoice (though we
-					// should be able to extend it further as BOLT 11 does have feature
-					// flags for the last hop node itself).
-					assert!(res.last().unwrap().pubkey == *target);
+
+					// Means we succesfully traversed from the payer to the payee, now
+					// save this path for the payment route. Also, update the liquidity
+					// remaining on the used hops, so that we take them into account
+					// while looking for more paths.
+					if ordered_hops.last().unwrap().route_hop.pubkey == *payee {
+						break 'path_walk;
+					}
+
+					new_entry = match dist.remove(&ordered_hops.last().unwrap().route_hop.pubkey) {
+						Some(payment_hop) => payment_hop,
+						// We can't arrive at None because of how we traverse with add_entry.
+						None => unreachable!(),
+					};
+					// We "propagate" the fees one hop backward (topologically) here,
+					// so that fees paid for a HTLC forwarding on the current channel are
+					// associated with the previous channel (where they will be subtracted).
+					ordered_hops.last_mut().unwrap().route_hop.fee_msat = new_entry.hop_use_fee_msat;
+					ordered_hops.last_mut().unwrap().route_hop.cltv_expiry_delta = new_entry.route_hop.cltv_expiry_delta;
+					ordered_hops.push(new_entry.clone());
 				}
-				if res.last().unwrap().pubkey == *target {
-					break;
+				ordered_hops.last_mut().unwrap().route_hop.fee_msat = value_contribution_msat;
+				ordered_hops.last_mut().unwrap().hop_use_fee_msat = 0;
+				ordered_hops.last_mut().unwrap().route_hop.cltv_expiry_delta = final_cltv;
+
+				let mut payment_path = PaymentPath {hops: ordered_hops};
+
+				// We could have possibly constructed a slightly inconsistent path: since we reduce
+				// value being transferred all the way, we could have violated htlc_minimum_msat on
+				// some channels we already passed (assuming dest->source direction). Here, we
+				// recompute the fees again, so that if that's the case, we match the currently
+				// underpaid htlc_minimum_msat with fees.
+				payment_path.update_value_and_recompute_fees(value_contribution_msat);
```

Comment:
Is the following method comment about `update_value_and_recompute_fees` still holds ? "Also, it's only safe to reduce the value, to not violate channel upper bounds")

Reply:
Yeah pretty sure it still holds. Adding a little clarification there.